### DISCUSSION TITLE
How do I customize the pattern of an antenna?

### DISCUSSION THREAD

**Discussion by vcyycv0** [NVlabs/sionna#409] [General]
For the "sionna.rt.antenna" function, I can't seem to customize the pattern of the antenna.
who can give me an example of importing the antenna pattern？

**Comment by faycalaa** [comment]
Hi,

You could have a look at [this repository](https://github.com/NVlabs/diff-rt-calibration/tree/main), in which a [custom antenna pattern is defined](https://github.com/NVlabs/diff-rt-calibration/blob/main/code/trainable_antenna_pattern.py), and then [used instead of one of the patterns available in Sionna](https://github.com/NVlabs/diff-rt-calibration/blob/main/notebooks/Synthetic_Data.ipynb).

**Reply by vcyycv0** [reply]
Thank you very much for your answer. now, I know how to design an antenna pattern. but I still have two questions. As the code reads：
Input
    -----
    theta : [array_like], tf.float
        Zenith angle

    phi = : [array_like], tf.float
        Azimuth angle

are theta and phi two-dimensional matrix that Represent the far-field spherical gain in both polarization directions? 

 c = tf.complex(tf.sqrt(gain), tf.zeros_like(gain))

What C stands for? and  What does C refer to in electromagnetism？

**Reply by faycalaa** [reply]
The `theta` and `phi` inputs are arrays (of any shape, not necessary matrices), that provide the direction `(theta, phi)` for which the antenna pattern is to be evaluated.

Check out the corresponding section of the [Primer on Electromagnetism](https://nvlabs.github.io/sionna/em_primer.html#far-field-of-a-transmitting-antenna) for details what is implemented.
`c` corresponds to the antenna pattern of a vertically polarized antenna pattern, i.e., $C = C_{\theta}$ and $C_{\phi} = 0$, in the primer.
The functions `polarization_model_1` and `polarization_model_2` then implement the polarization model from [3GPP TR 38.901](https://portal.3gpp.org/desktopmodules/Specifications/SpecificationDetails.aspx?specificationId=3173) (Section 7.3.2).

**Reply by vcyycv0** [reply]
Thank you for your answer. I have resolved  this problem. 
 In order to import the antenna pattern from the  HFSS into Sionna, I export the far-field patterns(two two-dimensional matrices  ,named pattern_theta and pattern_phi) in both polarization directions from HFSS. then， Use the following code as the return value for the pattern function.

```py     
    indices = tf.stack([theta_index, phi_index], axis=-1)
    gain_theta = tf.gather_nd(pattern_theta, indices)
    gain_phi = tf.gather_nd(pattern_phi, indices)
    #theta_index, phi_index are the pattern matrix size that I want to return
    return tf.cast(gain_theta, dtype), tf.cast(gain_phi, dtype)
```

  Thanks again for your reply.

**Reply by FFLMar** [reply]
Hi @vcyycv0 , I'm facing the same problem as you: trying to import and set a custom antenna pattern for the TX from outside (I have a gain matrice that is theta X phi). Could you share ore details about how you went about solving the issue. I can't figure how to define the antenna pattern and assign it to the TX. 
Thank you.
